### ISSUE TITLE
I want to do transfer learning with YOLOv8, could you please refer to the specific practices

### ISSUE THREAD

**Issue by KaedeharaT** [ultralytics/ultralytics#10428]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am conducting object detection on architectural drawings, and have trained a model using a dataset based on drawings from Company A, which performs well on other drawings from the same company. However, when using this model to predict on drawings from Company B, the detection rate is very low. The only difference between companies A and B is their drawing styles. I would like to perform transfer learning on the dataset based on Company A to improve its detection rate on Company B’s drawings. I am currently planning to use frozen layers, domain adaptation, and zero-shot transfer learning for this task.
Any suggestions would be appreciated.
![image](https://github.com/ultralytics/ultralytics/assets/110965108/82eb36f2-9bba-4287-a29d-6d82777a3a44)（Company A）
![image](https://github.com/ultralytics/ultralytics/assets/110965108/2abf33e3-6595-48f7-80f6-765267041923) （Company B）

**Comment by glenn-jocher** [comment]
@KaedeharaT certainly! For transfer learning with YOLOv8, leveraging the idea of starting from a pre-trained model and further training it on your targeted dataset (Company B's drawings) is a promising approach. Given the discrepancies in styles between Company A and B, it's essential to adapt the model for better performance.

Here's a brief step-by-step guide on how to proceed:

1. **Start with a Pre-trained Model:** Begin by loading a YOLOv8 model that has been pre-trained on a large dataset. This helps in having a good feature extraction base.
   ```python
   from ultralytics import YOLO
   model = YOLO('yolov8n.pt')  # Load YOLOv8n or another suitable pre-trained model
   ```

2. **Partial Freezing (Optional):** If you're planning on freezing layers, you can freeze all but the last few layers, allowing the final layers to learn features specific to Company B's styles.
   ```python
   model.freeze()  # Example method to freeze layers, adjust according to actual API
   ```

3. **Fine-tune on Company B's Dataset:** Train the model on your dataset comprising drawings from Company B. Adjust the epochs, batch size, and other hyperparameters based on your dataset size and the computational resources at your disposal.
   ```python
   results = model.train(data='companyB_data.yaml', epochs=30)
   ```

4. **Monitor and Adjust:** Keep an eye on the validation metrics. If the transfer learning process is not yielding the expected improvements, consider unfreezing more layers or tweaking other training hyperparameters.

5. **Domain Adaptation & Zero-Shot Learning:** These are more advanced strategies that might require custom implementation or adjustments. Domain adaptation could involve training with mixed datasets from both companies or employing techniques to minimize domain shifts. For zero-shot learning, adding textual descriptors or leveraging a dataset with broader classes might help.

The key to successful transfer learning with YOLOv8 is experimentation and iterative refinement based on performance metrics. Since each dataset and task is unique, the optimal settings and strategies will vary. Good luck with improving your model's detection rate for Company B's architectural drawings! 🚀

**Comment by KaedeharaT** [comment]
After completing training on a dataset based on Company A's drawings and generating a best.pt file, I then loaded this file to train on Company B's dataset (which has a few annotations). During training, I tried freezing 5, 10, and 15 layers to compare the results, but none were ideal. If too few layers are frozen, the new best.pt file significantly reduces the recognition rate on Company A's drawings, which I think might be due to too many layers undergoing parameter changes during training. On the other hand, if too many layers are frozen, the new best.pt file produces many erroneous detections for both companies (e.g., a round table in the drawing is mistakenly detected as a baselinemark). I trained it like this: 

**yolo detect train data=D:\dynamoFile\PythonScript\generalization\dataset\data.yaml model=D:/dynamoFile/PythonScript/runs/detect/train14/weights/best.pt epochs=100 imgsz=1000 device=0 freeze=5,** 

where train14 is the model trained from Company A's drawings.
Could you tell me where my approach might be going wrong? Is changing the learning rate key? I am planning to try that now.

**Comment by KaedeharaT** [comment]
I'm sorry for the late reply.Thank you for your guidance, I'll start trying it out now.

**Comment by glenn-jocher** [comment]
No worries at all! Best of luck with your experiments — I hope the adjustments work out well for you! If you hit any snags or have more questions, feel free to ask. Happy coding! 😊🚀

**Comment by KaedeharaT** [comment]
Hello, I have more question for which I would appreciate your assistance.
In the experiment, after training an object recognition model using Company A's dataset (which is extensively annotated), I used the layer freezing method to train on Company B's dataset (which has fewer annotations). **However, the PR curve from the training does not display information for 'aircon', yet during the prediction phase, 'aircon' does appear in the predictions.** Moreover, while annotating Company B's dataset, the number of annotations for 'aircon' was even greater than for 'window'. **When training on Company B's dataset, I used the best.pt file trained by Company A, rather than yolov8n.pt. Is this approach correct?**
![image](https://github.com/ultralytics/ultralytics/assets/110965108/b53eabf8-b1ae-4f07-9531-2751e75b4adc)
Fig1. This is the PR curve chart trained using Company A's dataset.
![image](https://github.com/ultralytics/ultralytics/assets/110965108/5f31f0a8-6a4a-4c04-a358-5eaf7467a9fc)
Fig2. This is the PR curve chart trained using Company B's dataset.（freeze=18，lr0=0.001，momentum=0.9，Optimizer=AdamW）
![image](https://github.com/ultralytics/ultralytics/assets/110965108/aba12cfa-f0e5-4a1e-9020-dd0ccfec644f)
Fig3. This image shows the prediction results of the model trained from Figure 2 on Company B's drawings.

**Comment by glenn-jocher** [comment]
@KaedeharaT hello! Thanks for sharing your training approach and results. 😊 It's quite intriguing that 'aircon' appears during predictions but not on the PR curve of the trained model.

1. **Using `best.pt` from Company A's Model**: Your approach to use `best.pt` from Company A's dataset as a start for training on Company B's dataset is generally sound, especially for transfer learning scenarios. This method leverages the pre-trained weights that potentially help the model to better generalize on different but related datasets.

2. **PR Curve Issue**: If 'aircon' is missing from the PR curve but appears in predictions, it could be related to how annotations were processed or a discrepancy in the class labels between training and validation phases. Ensure that:
   - The class 'aircon' is consistently labeled in both Company A's and B's datasets.
   - When merging datasets, the total count and labeling consistency remain intact.

Here’s a quick check you can perform to ensure the class labels are consistent:
```python
# Assuming you have a DataLoader for Company B
classes = model.names  # Class names as understood by the model
print("Classes detected by model:", classes)

# Check if 'aircon' is a recognized class
aircon_index = classes.index('aircon') if 'aircon' in classes else None
print("'aircon' class index:", aircon_index)
```

This should help clarify if the model has 'aircon' properly configured as a class during training. Keep an eye on the data loading and label processing steps! Let me know if this helps or if you have more questions! 🌟

**Comment by KaedeharaT** [comment]
Thank you for your kind support.
Regarding the issue of the 'aircon' category not appearing in the PR curve, I checked my dataset and found that it might be because I forgot to annotate 'aircon' in the validation set, though I did annotate it in the training and test sets. (This is because I wanted to generalize the YOLO model, which is trained with the well-annotated dataset from Company A, to the minimally annotated dataset from Company B.) Please let me know if my assumption is correct. Below are the PR curve and the prediction results of the transferred model:
![image](https://github.com/ultralytics/ultralytics/assets/110965108/ade01afe-251e-4bb8-9080-a547471cd88b)
Fig1. The PR Curve After TL
![image](https://github.com/ultralytics/ultralytics/assets/110965108/8a7e0272-2edb-4876-a06c-6e94b6be59fb)
Fig2. The prediction results


However, I tried the code you provided, and I wrote it like this:

```
from ultralytics import YOLO

if __name__ == '__main__':
    model = YOLO("D:/dynamoFile/PythonScript/runs/detect/train/weights/best.pt")

classes = model.names  # Class names as understood by the model
print("Classes detected by model:", classes)

aircon_index = classes.index('aircon') if 'aircon' in classes else None
print("'aircon' class index:", aircon_index)

```
However, I still got this result. Could you please tell me what's wrong with my code? (I'm a programming novice, sorry for the trouble.)
![image](https://github.com/ultralytics/ultralytics/assets/110965108/318cce52-f91f-4cc4-bb99-55fc68028265)

**Comment by glenn-jocher** [comment]
Hey, great job on checking the annotation status in different sets! 👍 If 'aircon' is not annotated in the validation dataset, it indeed won't appear in the PR curve. Your assumption is correct; the PR curves are derived from the validation set, so classes not represented there will not be displayed. 

Regarding the code snippet you've shared, it looks almost correct, but let's ensure the model is properly loaded and the classes are called within the right scope. Here’s a slightly adjusted version:

```python
from ultralytics import YOLO

if __name__ == '__main__':
    model = YOLO("D:/dynamoFile/PythonScript/runs/detect/train/weights/best.pt")
    classes = model.names  # Class names as understood by the model
    print("Classes detected by model:", classes)

    if 'aircon' in classes:
        aircon_index = classes.index('aircon')
        print("'aircon' class index:", aircon_index)
    else:
        print("'aircon' is not in the class list")
```

This slight adjustment ensures that any class-related operations are safely done after checking its presence in the list, which avoids potential errors if 'aircon' is not found.

If 'aircon' is still missing, make sure the same annotated classes for 'aircon' used during training are accessible to this script. If issues persist, it's possibly a matter of how the dataset was split or the model was trained. Happy coding! 😊

**Comment by glenn-jocher** [comment]
You're welcome! I'm glad to hear that you found the information helpful. 😊 Feel free to reach out anytime you have more questions or need further assistance. Happy experimenting! 🚀

**Comment by kanis777** [comment]
> Hey there! 🌟 It looks like you're on the right path with experimenting on layer freezing to achieve the right balance for transfer learning between Company A and B's datasets. Freezing layers can indeed be tricky; finding the sweet spot requires a bit of tinkering.
> 
> Your intuition about changing the learning rate is definitely worth exploring! A too high learning rate might cause overfitting to Company B's dataset, reducing performance on Company A's drawings. Conversely, a very low learning rate can slow down the adaptation or lead to underfitting. Here’s a quick suggestion to experiment with:
> 
> * Try a **gradual unfreezing approach**: Start training by freezing a larger portion of the layers, and gradually unfreeze layers in stages, allowing the model more flexibility to adapt as training progresses.
> * **Adjust the learning rate dynamically**: Start with a lower learning rate when more layers are frozen, and increase it as you unfreeze more layers. This can help in adapting the model gently to the new dataset.
> 
> Example snippet:
> 
> ```python
> # Assuming 'freeze' and 'lr' can be dynamically adjusted (this is a conceptual example)
> model.train(data='path/to/companyB_data.yaml', epochs=100, imgsz=1000, device=0, freeze=15, lr=0.001)
> ```
> 
> Remember, the key is in experimentation and slowly tweaking parameters based on your observations. Good luck! 🚀

Hey @glenn-jocher 
so after freezing the layers - should i unfreeze inorder to get the original learning or will the new model have that detail
?

**Comment by glenn-jocher** [comment]
Hi @kanis777,

Great question! When you freeze layers during transfer learning, the model retains the learned features from the original dataset (Company A) in those layers. If you unfreeze these layers later, the model can fine-tune these features based on the new dataset (Company B), potentially improving performance on the new data while still leveraging the original knowledge.

Here’s a suggested approach:

1. **Start with Frozen Layers**: Begin by freezing a significant portion of the layers to retain the learned features from Company A's dataset.
2. **Gradually Unfreeze**: As training progresses, gradually unfreeze layers to allow the model to adapt more to Company B's dataset. This helps in fine-tuning the model without losing the original learned features.

Example snippet for gradual unfreezing:
```python
# Initial training with frozen layers
model.train(data='path/to/companyB_data.yaml', epochs=50, imgsz=1000, device=0, freeze=15, lr=0.001)

# Gradually unfreeze layers and continue training
model.train(data='path/to/companyB_data.yaml', epochs=50, imgsz=1000, device=0, freeze=10, lr=0.0005)
```

This approach allows the model to first leverage the robust features learned from Company A and then gradually adapt to the nuances of Company B's dataset.

Remember, the key is to monitor the performance and adjust the parameters as needed. Happy experimenting! 🚀

**Comment by JuanDanielEspinoza** [comment]
Hi @glenn-jocher,
I’ve been reading about transfer learning in YOLOv11, and I’ve seen some comments suggesting that to retrain the model with frozen layers, you need to use a dataset that combines the original one with your new data. Is that actually necessary, or can I train the model using only my new dataset?

**Comment by glenn-jocher** [comment]
@JuanDanielEspinoza You don’t need to combine datasets—fine-tuning YOLO11 on only your new dataset is standard; mix in a small subset of the original data only if you must preserve performance on the old domain (to reduce forgetting).

```bash
yolo detect train model=yolo11n.pt data=companyB.yaml epochs=100 freeze=10
```

If your class list differs from the original, just ensure `companyB.yaml` defines the desired classes; metrics will reflect what’s in your `val` set. For context, see the Ultralytics overview of [transfer learning](https://www.ultralytics.com/glossary/transfer-learning) and our YOLO11 [training tips on using pre-trained weights](https://docs.ultralytics.com/guides/model-training-tips/).
